### ISSUE TITLE
Support INFOPLIST_PREPROCESS and related build settings

### ISSUE THREAD

**Issue by kphillisjr** [facebook/xcbuild#77]
I noticed that some of the property list files being built by xcbuild are not formatted correctly. I found a few example property lists that show generic examples that most developers will try to use in preprocessing of these files. In general this preprocessor works as a C Preprocessor with a few quirks added. Some notes on this can be found on Apple [Technical Note TN2175](https://developer.apple.com/library/prerelease/content/technotes/tn2175/_index.html). 

In general I would assume that the plist utility will support the following operations...
- Pre-processor Conditionals:  #if, #ifdef, #ifndef, #else, #elif and #endif.
- C-style Relational Operators ( see: [Wikipedia: Operators in C and C++](https://en.wikipedia.org/wiki/Operators_in_C_and_C%2B%2B)  )
- C-style Logical Operators ( see: [Wikipedia: Operators in C and C++](https://en.wikipedia.org/wiki/Operators_in_C_and_C%2B%2B)  )
- C-Style comments ``` /* Comment */```
- C++-Style comments ``` // Comment ``` ( when -traditional is not set )

Example 1: IOAudioFamily version 204.3
This is the one of the simplest examples of plist editing. All it does is define a few variables in an xcode configuration file and output it in the kernel extensions PLIST file..

Example 2: IOHIDFamily version 701.20.10
A slightly more complex example. This configures multiple property lists.

Example 3: IOStorageFamily version 191
This includes an example of preprocessor events to remove sections of the property list on a Embedded platform where TARGET_OS_EMBEDDED is defined in the TargetConditionals header.

Example 4: Some more Pre-processor examples.
For the most part if it's in the TargetConditionals header for OSX, it'll appear in the Property list utility.
 
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>PreProcess_KEYS</key>
	<dict>
		<key>IsEmbedded</key>
#if !TARGET_OS_EMBEDDED
		<true/>
#else 
		<false/>
#endif

		<key>IsMac</key>
#if TARGET_OS_MAC
		<true/>
#else 
		<false/>
#endif
		<key>IsPowerPC</key>
#if TARGET_CPU_PPC
		<true/>
#else 
		<false/>
#endif
	</dict>
</dict>
```

**Comment by grp** [comment]
This is the build setting `INFOPLIST_PREPROCESS`. You're right that it's not supported — there's [a TODO](https://github.com/facebook/xcbuild/blob/master/Libraries/pbxbuild/Sources/Phase/ProductTypeResolver.cpp#L309) for it.

If you build a project that uses this with Xcode, could you paste the commands it runs? That would help in figuring out the right way to implement it. It's a little different than most commands in that there's presumably two phases: `cpp` to preprocess, then `builtin-infoPlistUtility` as usual, and I'm not sure how the paths work out for that.

**Comment by grp** [comment]
Note that `bulitin-infoPlistUtility` is supported — that expands build setting constructs like this:

    <?xml version="1.0" encoding="UTF-8"?>
    <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
    <plist version="1.0">
    <dict>
        <key>CFBundleIdentifier</key>
        <string>com.yourcompany.$(PRODUCT_NAME:rfc1034identifier)</string>
        [...]
    </dict>

It also does a few other things, like adding extra keys based on the build environment and combining a few build settings.

**Comment by kphillisjr** [comment]
The basic syntax for "builtin-infoPlistUtility" is:
```sh
builtin-infoPlistUtility InputPropertylist.plist -expandbuildsettings -platform macosx -o Output.Plist
```
It's also good to note that this also accepts the argument ```-format``` where the next argument is mapped by the property lists ``INFOPLIST_OUTPUT_FORMAT``` key. This key has strict behavior of...

- when set to ``binary`` it causes a binary property list to be generated.
- when set to ``XML`` it causes a xml property list to be generated. ( The command line argument on this is ```-format xml```.
- when the value is not defined an xml property list is created.


also as a quick note,  ```INFOPLIST_OTHER_PREPROCESSOR_FLAGS`` is passed to the cc compiler. as far as commands go, with variables resolved, i believe this involves two stages...

```sh
#Stage 1: Run the property list through the c compiler.
Preprocess $TEMP_DIR/Preprocessed-Info.plist $INFOPLIST_FILE
    cd $PROJECT_DIR
    $TOOLCHAIN_DIR/usr/bin/cc -E -P -x c -Wno-trigraphs $PROJECT_DIR/$INFOPLIST_FILE -F$SYMROOT -isysroot $SDKROOT $INFOPLIST_OTHER_PREPROCESSOR_FLAGS -o $TEMP_DIR/Preprocessed-Info.plist

#Stage 2: Run the property list through the built in property list resolver.
ProcessInfoPlistFile $SYMROOT/$INFOPLIST_PATH $TEMP_DIR/Preprocessed-Info.plist
    cd $PROJECT_DIR
	builtin-infoPlistUtility $TEMP_DIR/Preprocessed-Info.plist -expandbuildsettings -format xml -platform macosx -o $SYMROOT/$INFOPLIST_PATH
```

**Comment by kphillisjr** [comment]
After some thought, I believe the builtin-infoPlistUtility simply handles all of the stuff in the ```${ /* Insert Variable here*/}``` or ```$(/* Insert variable here */ )``` strings.  For an idea on how these transformations work you can look at the [Code Worksho](http://codeworkshop.net/)p blog entry entitled [Xcode Build Setting Transformations](http://codeworkshop.net/posts/xcode-build-setting-transformations)

**Comment by grp** [comment]
Yeah, the `builtin-infoPlistUtility` transformations are already implemented in xcbuild. I'll take a look at implementing the preprocessing part.

**Comment by kphillisjr** [comment]
The preprocessor portion is straight forward. Just use gcc or clang/llvm to handle all the pre-processor steps. This is what apple did to accomplish this, and each compiler suite's documentation should help.

For GCC: [Using the GNU Compiler Collection (GCC): Preprocessor Options](https://gcc.gnu.org/onlinedocs/gcc/Preprocessor-Options.html)
For Clang/LLVM: [Clang Command guide man page](http://clang.llvm.org/docs/CommandGuide/clang.html) - This may not hold all the options, but in general all arguments to gcc should work on clang.

**Comment by grp** [comment]
I don't have a project handy to test with, but I think #78 should implement this.

**Comment by kphillisjr** [comment]
That should help, and IOStorageFamily Version 191 from [Apple's Open Source website](http://opensource.apple.com/) uses the c pre-processor steps in it's info.plist file. You can download it from the OS X 10.11.2 product page, and this should work as an example project to help.

**Comment by kphillisjr** [comment]
I noticed that ```INFOPLIST_PREPROCESSOR_DEFINITIONS``` is not included. I know of 5 different example cases that I have see with this... 

Example 1: an empty string.
```
INFOPLIST_PREPROCESSOR_DEFINITIONS=""
```

Example 2: A single file definition without -D.
```
INFOPLIST_PREPROCESSOR_DEFINITIONS="MY_DEFINE=1"
```
Example 3: An Array of defines.
```
INFOPLIST_PREPROCESSOR_DEFINITIONS = (
"MY_DEF_A=0",
"MY_DEF_B=1",
);
```

Example 4: An array of Defines with variable resolution.
```
INFOPLIST_PREPROCESSOR_DEFINITIONS = (
"MY_DEF_A=0",
"MY_DEF_B=1",
"MY_DEF_C=A${MY_DEF_A}B${MY_DEF_B}",
);
```

Example 5: An array of Random libraries/arguments  - Example is from LLVM's LLDB project file ( lldb.xcodeproj/project.pbxproj )
```
INFOPLIST_PREPROCESSOR_DEFINITIONS = (
	"-lobjc",
	"-lpython",
	"-framework",
	DebugSymbols,
	"-framework",
	Carbon,
	"-framework",
	Foundation,
	"-framework",
	AppKit,
	"-v",
	"-t",
	"-Wl,-v",
);
```

**Comment by grp** [comment]
`INFOPLIST_PREPROCESSOR_DEFINITIONS` should be included with the patch in #78, see this line: https://github.com/facebook/xcbuild/pull/78/files#diff-d14607be27fd32ae05a552241c5c1c98R313
